polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -1175,12 +1175,27 @@ Value MLIRScanner::SYCLCommonFieldLookup(Value V, size_t FNum,
   auto SYCLElemTy = cast<T>(ElemTy);
   assert(FNum < SYCLElemTy.getBody().size() && "ERROR");
 
-  const auto ElementType = SYCLElemTy.getBody()[FNum];
-  const auto ResultType = MemRefType::get(
-      Shape, ElementType, MemRefLayoutAttrInterface(), MT.getMemorySpace());
+  const Type ElementType = SYCLElemTy.getBody()[FNum];
+  Type ResultType;
+  Value ResultVal;
+
+  if (auto PT = ElementType.dyn_cast<LLVM::LLVMArrayType>()) {
+    ResultType =
+        LLVM::LLVMPointerType::get(ElementType, MT.getMemorySpaceAsInt());
+    auto Ptr = Builder.create<polygeist::Memref2PointerOp>(
+        Loc, LLVM::LLVMPointerType::get(SYCLElemTy, MT.getMemorySpaceAsInt()),
+        V);
+    Value GEPIdx[] = {Builder.create<arith::ConstantIntOp>(Loc, 0, 32),
+                      Builder.create<arith::ConstantIntOp>(Loc, FNum, 32)};
+    ResultVal = Builder.create<LLVM::GEPOp>(Loc, ResultType, Ptr, GEPIdx);
+  } else {
+    ResultType = MemRefType::get(
+        Shape, ElementType, MemRefLayoutAttrInterface(), MT.getMemorySpace());
+    ResultVal = Builder.create<polygeist::SubIndexOp>(Loc, ResultType, V,
+                                                      getConstantIndex(FNum));
+  }
```

Comment:
I guess the issue here is that `LLVM::LLVMArrayType` cannot be stored in a `MemRef`. Instead of this, I propose having a `memref` instead of a `LLVMArrayType` in the `stream` body and changing the lowering to LLVM to make sure an array is generated instead of a pointer.